src/operation.js

```diff
@@ -10,53 +10,61 @@ import buildResponse from './response';
 import Validator from './validation';
 
 
-function getRetries(request) {
-    if (
-        includes(
-            [
-                'post',
-                'patch',
-                'put',
-                'delete',
-            ],
-            lowerCase(request.method),
-        )
-    ) {
-        // Mutations will be retried on explicit initiation,
-        // instead of implicitly retried
-        return 0;
-    }
+function isMutationOperation(request) {
+    return includes(
+        [
+            'post',
+            'patch',
+            'put',
+            'delete',
+        ],
+        lowerCase(request.method),
+    );
+}
 
-    return get(request, 'retries', 0);
+function isProxyError(openApiError) {
+    return includes(
+        [
+            501, // Service unable to handle request due to wong routing
+            503, // Service instance does not exist.
+        ],
+        openApiError.code,
+    );
 }
 
-function isErrorRetryable(error) {
+function shouldRetryError(request, error, attempt) {
     const openApiError = normalizeError(error);
+    const maxAttempts = get(request, 'retries', 0) + 1;
+    const maxProxyAttempts = get(request, 'proxyRetries', 0) + 1;
 
-    if (
-        includes(
-            [
-                'econnaborted',
-                'econnreset',
-            ],
-            lowerCase(openApiError.code),
-        )
-    ) {
-        // Client timeout/error, retry
+    if (isProxyError(openApiError)) {
+        if (attempt >= Math.max(maxAttempts, maxProxyAttempts)) {
+            return false;
+        }
+
+        // We always want to retry 501 and 503 errors returned by HAProxy
         return true;
     }
 
+    if (attempt >= maxAttempts || isMutationOperation(request)) {
+        // Mutations will be retried on explicit initiation,
+        // instead of implicitly retried
+        return false;
+    }
+
     if (
         includes(
             [
-                502,
-                503,
-                504,
+                // Client timeout/error, retry
+                'econnaborted',
+                'econnreset',
+                // 50x responses are retryable
+                '502',
```

Comment:
why is 502 and 504 a string here but on L27 it looks like `openApiError.code` is an `int` for 501 and 503?

Reply:
Looking at the history of this as well, it looks like they were integers before these changes too?

Reply:
Previously there were two block doing the same but one looking for strings and one looking for integers. I just unified both in one block

Reply:
It's an array though, so why are we now doing `includes['502']` instead of `includes[502]`? Where 502 is a string instead of an int?

Reply:
Because we are comparing with the result of `lowerCase(openApiError.code),`